### PR TITLE
Fea/nn graph/warmup amp config

### PR DESCRIPTION
**Pull request by strint**
- WarmUpLR
  - eager
  - graph
-  amp & nn.Graph.config
- flow.config

### CODE REVIEW

File: python/oneflow/nn/optimizer/warm_up_lr.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import math
+
+from .optimizer import Optimizer
+from .lr_scheduler import WarmUpLrScheduler
+
+
+class WarmUpLR(WarmUpLrScheduler):
+    def __init__(
+        self,
+        lrsch_or_optimizer,
```
Comment (strint) [inline python/oneflow/nn/optimizer/warm_up_lr.py:25]: https://pytorch.org/docs/master/generated/torch.optim.lr_scheduler.WarmUpLR.html#torch.optim.lr_scheduler.WarmUpLR

改到了和torch开发分支中一样的接口。但是是个torch的增强版本：可以组合一个普通的LrScheduler

File: oneflow/core/framework/multi_client_session_context.cpp [hunk 2]
```diff
@@ -81,7 +81,7 @@ Maybe<void> MultiClientSessionContext::TryInit(const ConfigProto& config_proto)
 
       // TODO(chengcheng, guoran): handle CollectiveBoxingExecutor for multi-runtime in
       // Multi-Client.
-      resource.clear_collective_boxing_conf();
+      // resource.clear_collective_boxing_conf();
```
Comment (strint) [inline oneflow/core/framework/multi_client_session_context.cpp:84]: 开关层没问题，但是是不是下层执行还是有问题

Comment (chengtbf) [reply oneflow/core/framework/multi_client_session_context.cpp:84]: 这个我忘记 Revert 了。。你把 82 - 84 删掉吧。。。。 😂

File: oneflow/core/framework/multi_client_session_context.cpp [hunk 3]
```diff
@@ -78,10 +78,6 @@ Maybe<void> MultiClientSessionContext::TryInit(const ConfigProto& config_proto)
       } else {
         resource.set_cpu_device_num(gpu_device_num);
       }
-
-      // TODO(chengcheng, guoran): handle CollectiveBoxingExecutor for multi-runtime in
-      // Multi-Client.
-      resource.clear_collective_boxing_conf();
```
Comment (strint) [inline oneflow/core/framework/multi_client_session_context.cpp:84]: 注释里面说的功能已经支持了对吧

File: python/oneflow/test/graph/test_optimization_conf.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        flow.boxing.nccl.allow_fuse_broadcast(True)
+        flow.boxing.nccl.allow_fuse_mixed_ops(True)
+        flow.boxing.nccl.enable_use_buffer_to_fuse_all_reduce(True)
+        flow.boxing.nccl.set_stream_num(3)
+        flow.boxing.nccl.enable_all_to_all(True)
+        flow.boxing.nccl.enable_use_compute_stream(True)
+
+        flow.backends.cudnn.set_reserved_mem_mbytes(1000)
+
+        flow.utils.load_library("")
```
Comment (strint) [inline python/oneflow/test/graph/test_optimization_conf.py:74]: config接口调用的例子

File: python/oneflow/test/graph/test_optimization_conf.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+        class CustomGraphSysConf(flow.nn.Graph):
+            def __init__(self):
+                super().__init__()
+                self.m = CustomModule()
+
+                # amp
+                self.config.enable_amp(True)
+                grad_scaler = flow.nn.graph.amp.GradScaler(3000, 2.0, 0.5, 1000)
+                self.set_grad_scaler(grad_scaler)
```
Comment (strint) [inline python/oneflow/test/graph/test_optimization_conf.py:84]: amp调用的例子

File: python/oneflow/test/graph/test_optimization_conf.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+                self.config.allow_fuse_model_update_ops(True)
+                self.config.allow_fuse_add_to_output(True)
+                self.config.allow_fuse_cast_scale(True)
+                self.config.set_gradient_accumulation_steps(100)
```
Comment (strint) [inline python/oneflow/test/graph/test_optimization_conf.py:89]: graph config的例子

File: python/oneflow/test/graph/test_graph_lr_scheduler.py [hunk 7]
```diff
@@ -140,14 +138,104 @@ def one_iter():
...
+def _sgd_cosine_linear_fn(parameters):
+    of_sgd = flow.optim.SGD(parameters, lr=0.001)
+    alpha = 0.5
+    steps = 10
+    cosine_annealing_lr = flow.optim.lr_scheduler.CosineAnnealingLR(
+        of_sgd, steps=steps, alpha=alpha
+    )
+    linear_warmup_cosine_lr = flow.optim.lr_scheduler.WarmUpLR(
+        cosine_annealing_lr, warmup_factor=0.5, warmup_iters=5, warmup_method="linear"
+    )
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_lr_scheduler.py:183]: 组合warmup_lr + 普通lr_sch

File: python/oneflow/test/graph/test_graph_lr_scheduler.py [hunk 8]
```diff
@@ -140,14 +138,104 @@ def one_iter():
...
+    return of_sgd, constant_warmup_cosine_lr
+
+
+def _sgd_constant_fn(parameters):
+    of_sgd = flow.optim.SGD(parameters, lr=0.001)
+    alpha = 0.5
+    steps = 10
+    constant_warmup_lr = flow.optim.lr_scheduler.WarmUpLR(
+        of_sgd, warmup_factor=0.5, warmup_iters=5, warmup_method="constant"
+    )
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_lr_scheduler.py:170]: 直接使用warmup_lr

File: python/oneflow/nn/optimizer/lr_scheduler.py [hunk 9]
```diff
@@ -76,3 +78,36 @@ def step(self):
             group["lr"] = self.last_lr[i]
             if self.verbose:
                 self.print_lr(i, self.last_lr[i])
+
+
+class WarmUpLrScheduler(LrScheduler):
```
Comment (strint) [inline python/oneflow/nn/optimizer/lr_scheduler.py:83]: WarmUpLrScheduler基类

File: python/oneflow/nn/graph/amp/grad_scaler.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+
+class GradScaler(object):
```
Comment (strint) [inline python/oneflow/nn/graph/amp/grad_scaler.py:18]: GradScaler，当前只是生成nn.Graph下配置的工具函数

File: python/oneflow/test/graph/test_graph_lr_scheduler.py [hunk 11]
```diff
@@ -140,14 +138,104 @@ def one_iter():
...
+
+    def test_graph_constant(test_case):
+        _test_linear_graph_train_with_lr_sch(
+            test_case, 21, flow.device("cuda"), _sgd_constant_fn
+        )
+        _test_linear_graph_train_with_lr_sch(
+            test_case, 21, flow.device("cpu"), _sgd_constant_fn
+        )
+
+    def test_graph_cosine_linear(test_case):
```
Comment (strint) [inline python/oneflow/test/graph/test_graph_lr_scheduler.py:224]: 测试 optimizer + warmup_lr_sch  + lr_sch的组合，eager和lazy的对齐

File: oneflow/core/framework/multi_client_session_context.cpp (hunk 3 above)
Comment (leaves-zwx) [reply oneflow/core/framework/multi_client_session_context.cpp:84]: 已经支持了

File: python/oneflow/nn/graph/amp/grad_scaler.py (hunk 10 above)
Comment (leaves-zwx) [reply python/oneflow/nn/graph/amp/grad_scaler.py:18]: 位置有点奇怪，GradScaler 当初这么设计的原因就是为了 eager 能与 pytorch 对齐，lazy 能复用 eager 的部分功能。而 pytorch 的 GradScaler 在 `torch.cuda.amp.GradScaler` 这个位置。

这里看看位置上是否也要与 pytorch 的对齐。

File: python/oneflow/framework/config_util.py [hunk 12]
```diff
@@ -200,6 +200,26 @@ def reserved_device_mem_mbyte(val):
     sess.config_proto.resource.reserved_device_mem_mbyte = val
 
 
+def api_enable_cudnn_fused_normalization_add_relu(val: bool) -> None:
+    """Whether enable cudnn_fused_normalization_add_relu.
+
+    Args:
+        val (bool): whether enable or not
+    """
+    return enable_if.unique([enable_cudnn_fused_normalization_add_relu, do_nothing])(
```
Comment (leaves-zwx) [inline python/oneflow/framework/config_util.py:209]: 这个 normal_mode  未来会删除吧？现在只是为了兼容 global_function 而保留。@chengtbf

File: python/oneflow/nn/graph/graph.py [hunk 13]
```diff
@@ -469,14 +490,17 @@ def _shallow_repr(self):
         return shallow_repr
 
 
-class GraphConfig(FunctionConfig):
+class GraphConfig(object):
```
Comment (leaves-zwx) [inline python/oneflow/nn/graph/graph.py:493]: GraphConfig 要不要搬到另外一个文件里面去，python/oneflow/nn/graph/config.py。

File: python/oneflow/nn/optimizer/adam.py [hunk 14]
```diff
@@ -146,7 +146,11 @@ def generate_conf_for_graph(self, train_conf, vars_conf):
         for param_group in self.param_groups:
             optimizer_conf = train_conf.mutable_optimizer_conf().Add()
 
-            lr = param_group["lr"]
+            lr = (
+                param_group["initial_lr"]
```
Comment (leaves-zwx) [inline python/oneflow/nn/optimizer/adam.py:150]: initial_lr 这个是哪个 optimizer 用到的参数？

File: python/oneflow/nn/optimizer/lr_scheduler.py [hunk 15]
```diff
@@ -67,7 +67,9 @@ def get_last_lr(self):
     def print_lr(self, group_idx, lr):
```
Comment (leaves-zwx) [inline python/oneflow/nn/optimizer/lr_scheduler.py:67]: print_lr 为什么需要传入一个 lr， 这有点奇怪

File: python/oneflow/nn/graph/amp/grad_scaler.py (hunk 10 above)
Comment (strint) [reply python/oneflow/nn/graph/amp/grad_scaler.py:18]: 我们没有使用cuda的amp方案吧。
叫
```
flow.amp
```
是不是就可以

File: python/oneflow/nn/optimizer/lr_scheduler.py [hunk 16]
```diff
@@ -76,3 +78,36 @@ def step(self):
...
+    ):
+        self._lr_sch = None
+        isinstance(lrsch_or_optimizer, (LrScheduler, Optimizer))
+        if isinstance(lrsch_or_optimizer, LrScheduler):
+            self._lr_sch = lrsch_or_optimizer
+            for i, base_lr in enumerate(self._lr_sch.base_lrs):
+                # WarmupLRScheduler will restore lr changed by step() in self._lr_sch.__init___()
+                self._lr_sch._optimizer.param_groups[i][
+                    "lr"
+                ] = self._lr_sch._optimizer.param_groups[i]["initial_lr"]
```
Comment (leaves-zwx) [inline python/oneflow/nn/optimizer/lr_scheduler.py:95]: 或者这里改成 

```
self._lr_sch._optimizer.param_groups[i]["lr"] = base_lr
```

File: python/oneflow/nn/graph/graph.py (hunk 13 above)
Comment (strint) [reply python/oneflow/nn/graph/graph.py:493]: 可以，我单独开个文件

File: python/oneflow/nn/optimizer/adam.py (hunk 14 above)
Comment (strint) [reply python/oneflow/nn/optimizer/adam.py:150]: 是 lr_scheduler 创建的

File: python/oneflow/nn/optimizer/lr_scheduler.py [hunk 17]
```diff
@@ -76,3 +78,36 @@ def step(self):
...
+            for (i, group) in enumerate(self._optimizer.param_groups):
+                group["lr"] = self.last_lr[i]
+                if self.verbose:
+                    self.print_lr(i, self.last_lr[i])
+        elif self.last_step == self.warmup_iters:
+            self._lr_sch.last_step = self.last_step - 1
+            self._lr_sch.last_lr = self.last_lr
+            self._lr_sch.step()
+        else:
+            self._lr_sch.step()
```
Comment (leaves-zwx) [inline python/oneflow/nn/optimizer/lr_scheduler.py:113]: 假如 warmup_iters == 0，在调用 warmup_lr.step() 的时候，_lr_sch.step() 其实已经被调了两次，且之后所有 step 调用，warmup_lr 的 step 总比 base_lr_sch 的 step 大 1

File: python/oneflow/nn/optimizer/lr_scheduler.py (hunk 16 above)
Comment (leaves-zwx) [reply python/oneflow/nn/optimizer/lr_scheduler.py:95]: 下面基类的 init 函数里面不是又会把 optimizer.param_groups[i]["lr"] 修改掉吗？

Comment (wyg1997) [reply python/oneflow/nn/optimizer/lr_scheduler.py:95]: 下面改掉的话也是由外层这个 scheduler 重写的 get_lr() 控制的，这时候写改掉就没问题了

File: python/oneflow/nn/optimizer/adam.py (hunk 14 above)
Comment (wyg1997) [reply python/oneflow/nn/optimizer/adam.py:150]: 这里的设计是如果Optimizer 被 LrScheduler 包起来的话，LrScheduler 会对 Optimizer 传入的初始 lr 做一个备份，有些 LrScheduler 的更新会用到这个初始的 lr

File: python/oneflow/framework/config_util.py (hunk 12 above)
Comment (strint) [reply python/oneflow/framework/config_util.py:209]: 嗯，应该后面做一次统一的清理。这里为了保持风格一致，专门加上的。

File: python/oneflow/nn/optimizer/lr_scheduler.py [hunk 18]
```diff
@@ -76,3 +78,36 @@ def step(self):
...
             if self.verbose:
                 self.print_lr(i, self.last_lr[i])
+
+
+class WarmUpLrScheduler(LrScheduler):
+    def __init__(
+        self, lrsch_or_optimizer, last_step=-1, verbose=False,
+    ):
+        self._lr_sch = None
+        isinstance(lrsch_or_optimizer, (LrScheduler, Optimizer))
```
Comment (leaves-zwx) [inline python/oneflow/nn/optimizer/lr_scheduler.py:88]: assert

File: python/oneflow/nn/graph/graph.py [hunk 19]
```diff
@@ -64,6 +64,16 @@ def name(self):
     def training(self):
         return self.config.training
 
+    @property
+    def _config_proto(self):
+        return self.config.proto
+
+    @property
+    def _optimization_conf_proto(self):
```
Comment (chengtbf) [inline python/oneflow/nn/graph/graph.py:72]: 这个只是为了 debug 打印吧

**Review by chengtbf — COMMENTED** [review]
LR 部分的内容 @leaves-zwx 文骁 Review 吧，通过了就能合并~

File: python/oneflow/test/graph/test_optimization_conf.py [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+        flow.utils.load_library("")
+
+        class CustomGraphSysConf(flow.nn.Graph):
+            def __init__(self):
+                super().__init__()
+                self.m = CustomModule()
+
+                # amp
+                self.config.enable_amp(True)
+                grad_scaler = flow.amp.GradScaler(3000, 2.0, 0.5, 1000)
```
Comment (chengtbf) [inline python/oneflow/test/graph/test_optimization_conf.py:84]: 这里需要提供 key args 把，不然别人不知道  3000、2、1000 都是啥

**Review by chengtbf — APPROVED** [review]

**Review by leaves-zwx — APPROVED** [review]

File: python/oneflow/test/graph/test_optimization_conf.py (hunk 20 above)
Comment (strint) [reply python/oneflow/test/graph/test_optimization_conf.py:84]: 这里和torch接口对齐了，用户传参时也可以写 arg=val，也可以不写

Comment (chengtbf) [reply python/oneflow/test/graph/test_optimization_conf.py:84]: 对，但是我觉得我们的示例代码要写明白，参考：

```python
# Assuming optimizer uses lr = 0.05 for all groups
# lr = 0.025    if epoch == 0
# lr = 0.03125  if epoch == 1
# lr = 0.0375   if epoch == 2
# lr = 0.04375  if epoch == 3
# lr = 0.005    if epoch >= 4
scheduler = WarmUpLR(self.opt, warmup_factor=0.5, warmup_iters=4, warmup_method="linear")
for epoch in range(100):
    train(...)
    validate(...)
    scheduler.step()
```

Comment (chengtbf) [reply python/oneflow/test/graph/test_optimization_conf.py:84]: 我们的测试脚本，用户也会尝试模仿。所以最好还是带上参数更明白

Comment (strint) [reply python/oneflow/test/graph/test_optimization_conf.py]: 有道理，已经加上

File: python/oneflow/nn/graph/graph.py (hunk 19 above)
Comment (strint) [reply python/oneflow/nn/graph/graph.py:72]: 嗯，为了内部debug，加下滑线前缀的方法都是内部使用的私有方法，不保证稳定

File: python/oneflow/boxing/nccl/__init__.py [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+from oneflow.framework.config_util import api_nccl_num_streams as set_stream_num
+
+from oneflow.framework.config_util import (
+    api_nccl_enable_all_to_all as enable_all_to_all,
+)
+
+from oneflow.framework.config_util import (
+    api_nccl_use_compute_stream as enable_use_compute_stream,
```
Comment (chengtbf) [inline python/oneflow/boxing/nccl/__init__.py:35]: 这里不是有吗？ 😂   @strint @leaves-zwx 

```python
flow.boxing.nccl.enable_use_compute_stream()
```
